### ISSUE TITLE
msg.message.user is now returning Slack User objects, not Hubot User objects

### ISSUE THREAD

**Issue by mistydemeo** [slackapi/hubot-slack#414]
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/hubot-slack/blob/master/CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://github.com/slackapi/hubot-slack/blob/master/CODE_OF_CONDUCT.md).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

#### Description
Somewhere between 3.4.0 and 4.3.3, hubot-slack changed its behaviour for returning User objects. In 3.4.0 and earlier versions, the `msg.message.user`/`msg.envelope.user` object is an instance of Hubot's own User object, as stored in the brain (and potentially decorated by other scripts). In the current version, `@slack/client`'s User objects are being returned instead. Since these have different properties with different names, this has the potential to break scripts. I didn't see this documented in the [upgrade guide](https://slackapi.github.io/hubot-slack/upgrading#upgrading-from-version-3-or-earlier), so I thought this might have been an unintentional change.

#### Reproducible in:
{project_name} version: 4.3.3
OS version(s): Any

#### Steps to reproduce:

1. Access the `msg.message.user` object in a Slack script.

### Expected result:
User should be a Hubot User object.

### Actual result:
User is a Slack User object.

**Comment by mistydemeo** [comment]
I suspect this may have changed in 744359dbb92af9b9ac5053b1577ed44302c36cb6 (in release 4.0.0 and later, along with the unreleased 3.5.0). Prior to this commit, the user object is processed into a full hubot user here: https://github.com/slackapi/hubot-slack/commit/744359dbb92af9b9ac5053b1577ed44302c36cb6#diff-44aca805fd3e634b2ed4f6bc2ccb0749L146

```js
    # Process the user into a full hubot user
    user = @robot.brain.userForId msg.user
    user.room = channel.name
```

In 744359dbb92af9b9ac5053b1577ed44302c36cb6, which updated the code for `@slack/client` v3, that code was dropped and the user lookup was moved elsewhere - and turned into a `@client.dataStore.getUserById` call, which returns a Slack User object. Since there was no mention of this in the commit message, and no comment to this effect, I'm pretty sure this was not intentional.

cc @johnagan, who authored the above commit

**Comment by mistydemeo** [comment]
Just checking in - it'd be very helpful to get a response to this.

**Comment by aoberoi** [comment]
@mistydemeo is this still an issue? we no longer use `@client.dataStore` for the user data (although it exists there - see #385), but instead use the API you mentioned: https://github.com/slackapi/hubot-slack/blob/6ef2ac4a90c303948eb4b7427fca1de606d60deb/src/bot.coffee#L240

```
@robot.brain.userForId user.id, newUser
```

**Comment by mistydemeo** [comment]
This is still an issue, yes - the user object being returned is still a Slack User object, not a Hubot User object.

**Comment by aoberoi** [comment]
okay i understand. it looks like there are a couple places where we need to adjust the implementation to make this work:

- https://github.com/slackapi/hubot-slack/blob/6ef2ac4a90c303948eb4b7427fca1de606d60deb/src/bot.coffee#L158-L162
- https://github.com/slackapi/hubot-slack/blob/6ef2ac4a90c303948eb4b7427fca1de606d60deb/src/bot.coffee#L204-L205

**Comment by kevb** [comment]
Hi, possible to get an update on this? New hubot user here with broken hubot-auth integration due to this error. Not sure I want to dive in and fix myself without getting more familiar...

**Comment by nine-inch-rails** [comment]
A fix I've found is to use `robot.brain.userForName(msg.message.user.name)` in place of `msg.message.user`

**Comment by shanedewael** [comment]
Fixed with #472
